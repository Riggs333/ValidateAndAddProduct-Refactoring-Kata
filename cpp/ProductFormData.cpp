#include "ProductFormData.h"

ProductFormData::ProductFormData(std::string name, std::string type, float weight, float suggestedPrice,
                                 bool packagingRecyclable) : name(name), type(type), weight(weight),
                                 suggestedPrice(suggestedPrice), packagingRecyclable(packagingRecyclable) {

}
