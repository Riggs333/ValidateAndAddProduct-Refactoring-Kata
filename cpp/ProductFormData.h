#ifndef VALIDATE_AND_ADD_PRODUCT_REFACTORING_KATA_PRODUCTFORMDATA_H
#define VALIDATE_AND_ADD_PRODUCT_REFACTORING_KATA_PRODUCTFORMDATA_H


#include <string>

class ProductFormData {

public:
    ProductFormData(std::string name, std::string type, float weight, float suggestedPrice, bool packagingRecyclable);

    std::string name;
    std::string type;
    float weight;
    float suggestedPrice;
    bool packagingRecyclable;
};


#endif //VALIDATE_AND_ADD_PRODUCT_REFACTORING_KATA_PRODUCTFORMDATA_H
