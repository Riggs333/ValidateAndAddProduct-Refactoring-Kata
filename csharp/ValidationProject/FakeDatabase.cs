namespace Validation
{
    internal class FakeDatabase : IDatabaseAccess
    {
        public Product Product { get; set; }
        public int storeProduct(Product product)
        {
            Product = product;
            return 0;
        }
    }
}