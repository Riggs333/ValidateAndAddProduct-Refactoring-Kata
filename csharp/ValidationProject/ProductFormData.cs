namespace Validation
{
    internal class ProductFormData
    {
        public string Name { get; }
        public string Type { get; }
        public double Weight { get; }
        public double SuggestedPrice { get; }
        public bool PackagingRecyclable { get; }

        public ProductFormData(string name, string type, double weight, double suggestedPrice, bool packagingRecyclable)
        {
            Name = name;
            Type = type;
            Weight = weight;
            SuggestedPrice = suggestedPrice;
            PackagingRecyclable = packagingRecyclable;
        }
    }
}