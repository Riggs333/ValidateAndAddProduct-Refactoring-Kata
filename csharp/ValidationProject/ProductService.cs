namespace Validation
{
    internal class ProductService
    {
        private readonly IDatabaseAccess _db;

        public ProductService(IDatabaseAccess db)
        {
            _db = db;
        }

        public Response ValidateAndAdd(ProductFormData productData)
        {
            if ("" == (productData.Name))
            {
                return new Response(0, -2, "Missing Name");
            }

            if ("" == (productData.Type))
            {
                return new Response(0, -2, "Missing Type");
            }

            Product product = new Product(productData.Name);
            product.Type = ("Unknown");

            if ("Eyeshadow" == (productData.Type) || "Mascara" == (productData.Type))
            {
                product.Type = (productData.Type);
                product.Family = (ProductFamily.EYES);
                if ("Eyeshadow" == (productData.Type) && product.Name.Contains("Queen"))
                {
                    product.Range = (ProductRange.QUEEN);
                }
            }

            product.Range = (ProductRange.BUDGET);
            if (productData.PackagingRecyclable)
            {
                product.Range = (ProductRange.PROFESSIONAL);
            }

            if ("Foundation" == (productData.Type))
            {
                if (productData.SuggestedPrice > 10)
                {
                    product.Range = (ProductRange.PROFESSIONAL);
                }
            }

            if ("Lipstick" == (productData.Type))
            {
                product.Type = (productData.Type);
                product.Family = (ProductFamily.LIPS);
                if (productData.SuggestedPrice > 10)
                {
                    product.Range = (ProductRange.PROFESSIONAL);
                }

                if (productData.SuggestedPrice > 20)
                {
                    if (productData.Weight > 0 && productData.Weight < 10)
                    {
                        return new Response(0, -1, "Error - failed quality check for Queen Range");
                    }

                    product.Range = (ProductRange.QUEEN);
                }
            }

            if ("Mascara" == (productData.Type))
            {
                product.Family = (ProductFamily.LASHES);
                if (productData.SuggestedPrice > 15)
                {
                    product.Range = (ProductRange.PROFESSIONAL);
                }

                if (productData.SuggestedPrice > 25 && productData.PackagingRecyclable)
                {
                    product.Range = (ProductRange.QUEEN);
                }
            }

            if (productData.Weight < 0)
            {
                return new Response(0, -3, "Weight error");
            }

            product.Weight = (productData.Weight);

            if ("Blusher" == (productData.Type) || "Foundation" == (productData.Type))
            {
                product.Type = (productData.Type);
                product.Family = (ProductFamily.SKIN);
                if ("Blusher" == (productData.Type) && productData.Weight > 10)
                {
                    return new Response(0, -3, "Error - weight too high");
                }
            }

            if (!productData.PackagingRecyclable && product.Range == ProductRange.QUEEN)
            {
                return new Response(0, -1, "Error - failed quality check for Queen Range");
            }

            if ("Unknown" == (product.Type))
            {
                return new Response(0, -1, "Unknown product type " + productData.Type);
            }

            return new Response(_db.storeProduct(product), 0, "Product Successfully Added");
        }
    }
}