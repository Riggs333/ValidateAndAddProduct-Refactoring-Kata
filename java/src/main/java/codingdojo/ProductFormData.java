package codingdojo;

/**
 * Received from the client with all the details of the new product to be added to the catalogue.
 */
public class ProductFormData {
    private final String name;
    private String type;
    private final double weight;
    private final double suggestedPrice;
    private final boolean packagingRecyclable;

    public ProductFormData(String name, String type, double weight,
                           double suggestedPrice, boolean packagingRecyclable) {

        this.name = name;
        this.type = type;
        this.weight = weight;
        this.suggestedPrice = suggestedPrice;
        this.packagingRecyclable = packagingRecyclable;
    }

    public String getType() {
        return type;
    }

    public String getName() {
        return name;
    }

    public double getWeight() {
        return weight;
    }

    public double getSuggestedPrice() {
        return suggestedPrice;
    }

    public boolean  isPackagingRecyclable() {
        return packagingRecyclable;
    }
}
