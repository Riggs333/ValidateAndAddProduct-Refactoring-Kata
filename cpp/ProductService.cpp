#include "ProductService.h"
#include "ProductFormData.h"
#include "Product.h"

ProductService::ProductService(IDatabase * database) : db(database) {

}

Response ProductService::validateAndAdd(ProductFormData *productData) {
    if (productData->name.empty())
    {
        return Response(0, -2, "Missing Name");
    }

    if (productData->type.empty())
    {
        return Response(0, -2, "Missing Type");
    }

    auto product = new Product(productData->name);
    product->type = ("Unknown");

    if ("Eyeshadow" == (productData->type) || "Mascara" == (productData->type))
    {
        product->type = (productData->type);
        product->family = ProductFamily::EYES;
        if ("Eyeshadow" == (productData->type) && product->name.find("Queen") != std::string::npos)
        {
            product->range = (ProductRange::QUEEN);
        }
    }

    product->range = (ProductRange::BUDGET);
    if (productData->packagingRecyclable)
    {
        product->range = (ProductRange::PROFESSIONAL);
    }

    if ("Foundation" == (productData->type))
    {
        if (productData->suggestedPrice > 10)
        {
            product->range = (ProductRange::PROFESSIONAL);
        }
    }

    if ("Lipstick" == (productData->type))
    {
        product->type = (productData->type);
        product->family = (ProductFamily::LIPS);
        if (productData->suggestedPrice > 10)
        {
            product->range = (ProductRange::PROFESSIONAL);
        }

        if (productData->suggestedPrice > 20)
        {
            if (productData->weight > 0 && productData->weight < 10)
            {
                return Response(0, -1, "Error - failed quality check for Queen Range");
            }

            product->range = (ProductRange::QUEEN);
        }
    }

    if ("Mascara" == (productData->type))
    {
        product->family = (ProductFamily::LASHES);
        if (productData->suggestedPrice > 15)
        {
            product->range = (ProductRange::PROFESSIONAL);
        }

        if (productData->suggestedPrice > 25 && productData->packagingRecyclable)
        {
            product->range = (ProductRange::QUEEN);
        }
    }

    if (productData->weight < 0)
    {
        return Response(0, -3, "Weight error");
    }

    product->weight = (productData->weight);

    if ("Blusher" == (productData->type) || "Foundation" == (productData->type))
    {
        product->type = (productData->type);
        product->family = (ProductFamily::SKIN);
        if ("Blusher" == (productData->type) && productData->weight > 10)
        {
            return Response(0, -3, "Error - weight too high");
        }
    }

    if (!productData->packagingRecyclable && product->range == ProductRange::QUEEN)
    {
        return Response(0, -1, "Error - failed quality check for Queen Range");
    }

    if ("Unknown" == (product->type))
    {
        return Response(0, -1, "Unknown product type " + productData->type);
    }

    return Response(db->storeProduct(product), 0, "Product Successfully Added");

}
