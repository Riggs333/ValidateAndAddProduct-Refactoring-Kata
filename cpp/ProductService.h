#ifndef VALIDATE_AND_ADD_PRODUCT_REFACTORING_KATA_PRODUCTSERVICE_H
#define VALIDATE_AND_ADD_PRODUCT_REFACTORING_KATA_PRODUCTSERVICE_H


#include "Response.h"
#include "IDatabase.h"
#include "ProductFormData.h"

class ProductService {

public:
    explicit ProductService(IDatabase *database);

    Response validateAndAdd(ProductFormData *productData);

private:
    IDatabase *db;
};


#endif //VALIDATE_AND_ADD_PRODUCT_REFACTORING_KATA_PRODUCTSERVICE_H
