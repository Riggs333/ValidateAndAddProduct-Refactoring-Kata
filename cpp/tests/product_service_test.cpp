#include <ProductFormData.h>
#include <ProductService.h>
#include "doctest.h"
#include "ApprovalTests.hpp"
#include "FakeDatabase.h"

TEST_CASE ("test_validate_and_add") {
    auto productData = new ProductFormData("Sample product", "Lipstick", 5, 10, false);
    auto db = new FakeDatabase();
    auto sut = new ProductService(db);

    Response response = sut->validateAndAdd(productData);

    std::string responseAndProduct;
    responseAndProduct.append(response.to_string());
    responseAndProduct.append(" ");
    responseAndProduct.append(db->product->to_string());

    Approvals::verify(responseAndProduct);
}