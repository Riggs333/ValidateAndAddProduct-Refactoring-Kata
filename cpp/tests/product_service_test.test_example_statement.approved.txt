Response{productId=0, statusCode=0, message='Product Successfully Added'} Product{name='Sample product', type='Lipstick', weight=5.000000, family=1, range=1}
